### ISSUE TITLE
BDI and multi-threads

### ISSUE THREAD

**Issue by ptaillandier** [gama-platform/gama#1887]
### Steps to reproduce
1. Open the attached model  

2. test the batch experiment using the multi-threads option (default preference)

3.test the batch experiment without using the multi-threads option

[models.zip](https://github.com/gama-platform/gama/files/323009/models.zip)

### Expected behavior
The result should be similar

### Actual behavior
The result will be very different (not linked to the stochasticity of the model - see the nb_defending_D value). 

for example:
with multi-threads: *** TIME : 102973.0 nb_in_house_D:51.49367886357501 nb_escaping_D:48.50632113642499 nb_defending_D:0.0
without multi-threads: *** TIME : 59291.0 nb_in_house_D:59.9323845574436 nb_escaping_D:15.060873048296289 nb_defending_D:25.006742394260108

The good results are the ones generated without the multi-threads option.

### System and version 
Git version with both Mac OS and Windows

**Comment by AlexisDrogoul** [comment]
We should probably check if there are some problems of static variables / stateful instances of skills, etc. in the BDI plugin. As a matter of fact, the only difference between multi-thread / mono-thread is that the simulations run in parallel (obviously) so any shared state can cause a problem. Have you tried comparing non-batch simulations ? I.e. if you run 2 simulations, do they each give you the same result as if you run 1 ?

**Comment by AlexisDrogoul** [comment]
I've just pushed a fix for another issue -- and it might be related to this. Can you check ?

**Comment by ptaillandier** [comment]
Unfortunately, it is still not working.

2016-06-20 19:58 GMT+02:00 Alexis Drogoul <notifications@github.com>:

**Comment by AlexisDrogoul** [comment]
Yet another possible fix, removing a thread unsafe reference to a "result". No improvement ?

**Comment by ptaillandier** [comment]
Nope, it is still not working.

2016-06-21 11:50 GMT+02:00 Alexis Drogoul <notifications@github.com>:

**Comment by AlexisDrogoul** [comment]
So we probably have to go down to the level of Predicates, etc. to verify the thread safeness of everything. The good starting point is to try to understand, when you have two simulations running, which objects are shared between them (besides the statements, expressions, etc. which should, most of them, be thread safe because they are stateless). Or if there are some caches of objects somewhere..

**Comment by AlexisDrogoul** [comment]
I've begun to browse the classes in the simplebdi plugin and I am a bit lost. What are the classes that are actually functional ? It seems that neither BDIPlan nor SimpleBDIPlan (which is empty in my version) can work properly. As an example, the `getIntention()` method in BDIPlan is supposed to return a Predicate, but will probably throw a ClassCastException when doing so, as it casts an IExpression into a Predicate (which is not possible since Predicates are not Expressions).

So, which classes can be inspected for thread safeness and which should be forgotten (in that case, feel free to remove them, by the way) ? 

Other things (not related to the issue, I guess): 

- I noticed in some places (SimpleBDIArchitecture, lines 334 and 381) the use of `scope.getExperiment().getRandomGenerator()`. Be aware that in that case you will be using the RNG from the experiment, not the one from the simulation. So all simulations will use the same. 

- Code like:

```
				if ((GamaList) ((Predicate) statement).getOnHoldUntil() != null) {
					if (((GamaList) ((Predicate) statement).getOnHoldUntil()).contains(newPredicate)) {
						((GamaList) ((Predicate) statement).getOnHoldUntil()).remove(newPredicate);
					}
				}
				if ((ArrayList) ((Predicate) statement).getSubintentions() != null) {
					if (((ArrayList) ((Predicate) statement).getSubintentions()).contains(newPredicate)) {
						((ArrayList) ((Predicate) statement).getSubintentions()).remove(newPredicate);
					}
				}
```

is not only inefficient (how many casting are done per cycles ??) but also error-prone. Why not giving `getSubIntentions()` a proper type from the beginning and use this type ? Why not changing the type of `statement` immediately so that you dont have to worry about it after ? 

- You create a copy of the scope in `ConsciousContagionStatement` but never release it. It is maybe ok (never tested this myself), but it clearly opens the way for memory related errors. All unused scopes should be released using `GAMA.releaseScope(...)'.

**Comment by mathieuBourgais** [comment]
I've cleaned the code following your remarks. Now, all the classes are functional.
